lib/src/ruby/module.cc

```diff
@@ -1074,8 +1074,12 @@ namespace facter { namespace ruby {
         if (it == _facts.end()) {
             // Before adding the first fact, call facts to ensure the collection is populated
             facts();
-            it = _facts.insert(make_pair(fact_name, fact::create(name))).first;
-            ruby.rb_gc_register_address(&it->second);
+            // facts() may add entries to _facts, so check again to ensure entry has not already been added (avoid duplicate Ruby GC registration)
+            it = _facts.find(fact_name);
+            if (it == _facts.end()) {
```

Comment:
The slightly more optimized way to do this would be to use `std::map::upper_bound` instead of `find`. The result of `upper_bound` can be used as a hint for `insert` (it used to be that you used `lower_bound`, but apparently that changed in C++11). However, I don't think this section will be dealing with data sets large enough for that to matter.